Major fixes to s390x ABI and add TAILCALL trace support

monojenkins opened the pull request:
!! This PR is a copy of mono/mono#19611,  please do not edit or review it in this repo !!<br/>Do not automatically approve this PR:<br/><br/>* Consider how the changes affect configurations in this repo,<br/>* Check effects on files that are not mirrored,<br/>* Identify test cases that may be needed in this repo.<br/><br/>!! Merge the PR only after the original PR is merged !!<br/><hr/><br/>This pull request:
* Fixes the s390x ABI 
  * For the case where structures are passed as parameters on the stack
  * Fixes the ```OP_S390_MOVE``` definition so register scheduling works correctly
  * Optimizes ```localloc``` generation
  * Fixes tailcall processing 
    * Fixes register reloading
    * Adds ```OP_TAILLCALL_REG``` for s390x
    * Clarifies the conditions when tailcalls are supported on s390x
* Change function comment style to be doxygen-compatible 
* Remove dead code
* Extract hardware facilities from CPU to enable use of newer opcodes
* Adds altstack support to exception processing for s390x
* Fixes a crash when using ```--trace``` with programs using tailcalls by providing its own icall method:
  * ```mono_trace_tail_method``` will use the callee and caller methods to report on the tailcall. 
  * At the moment tailcalls are trace in ```mono_trace_leave_method``` where the second operand is expected to be a ```MonoProfileCallContext``` from which the return type and value are extracted which is meaningless for tail calls and worse may cause a SEGV.
  * The added call will use the parameters that are set up by ```mini_profiler_emit_tail_call```
* Fixes crashes in ```icall_System_Runtime_InteropServices_Marshal_copy_to_unmanaged``` and ```icall_System_Runtime_InteropServices_Marshal_copy_from_unmanaged```:
  * We now don't check the src/dest when the length field is zero so if either address is null it doesn't cause an exception
  * This is exhibited when an empty UTF16 string is converted to Byte[]. See this excerpt from ```WriteUTF16```in ```BlobBuilder.cs```:
    * For BigEndian systems the ```Encoding.Unicode.GetBytes(value);``` returns a null pointer for empty strings
    * ```WriteBytesUncheced``` invokes ```copy_from_unmanaged_fixed``` which crashes when given a null source pointer
```
            if (BitConverter.IsLittleEndian)
            {
                fixed (char* ptr = &value[0])
                {
                    WriteBytesUnchecked((byte*)ptr, value.Length * sizeof(char));
                }
            }
            else
            {
                byte[] bytes = Encoding.Unicode.GetBytes(value);
                fixed (byte* ptr = &bytes[0])
                {
                    WriteBytesUnchecked((byte*)ptr, bytes.Length);
                }
            }
```
With these changes s390x now works with the Roslyn compiler. The results of make check now show a higher pass rate:
```
22 test(s) passed, 0 test(s) did not pass.
617 test(s) passed, 71 test(s) did not pass.
216 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
9 test(s) passed, 0 test(s) did not pass.
9 test(s) passed, 0 test(s) did not pass.
4 test(s) passed, 0 test(s) did not pass.
4 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
6 test(s) passed, 0 test(s) did not pass.
Tests run: 10794, Passed: 10639, Errors: 2, Failures: 54, Inconclusive: 0
Tests run: 10, Passed: 10, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 634, Passed: 630, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 6401, Passed: 6260, Errors: 11, Failures: 6, Inconclusive: 0
Tests run: 2024, Passed: 1824, Errors: 0, Failures: 4, Inconclusive: 0
Tests run: 491, Passed: 466, Errors: 1, Failures: 0, Inconclusive: 0
Tests run: 200, Passed: 198, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1596, Passed: 1590, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 250, Passed: 250, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1821, Passed: 1804, Errors: 5, Failures: 2, Inconclusive: 0
Tests run: 24, Passed: 24, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 4, Passed: 4, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 89, Passed: 87, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 58, Passed: 58, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2405, Passed: 2054, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 100, Passed: 100, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 15, Passed: 15, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2891, Passed: 2774, Errors: 1, Failures: 6, Inconclusive: 0
Tests run: 22, Passed: 22, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 83, Passed: 33, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 61, Passed: 61, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 101, Passed: 97, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 4396, Passed: 4371, Errors: 1, Failures: 2, Inconclusive: 0
Tests run: 247, Passed: 244, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 256, Passed: 123, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 19, Passed: 19, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 5, Passed: 5, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 87, Passed: 26, Errors: 51, Failures: 7, Inconclusive: 0
Tests run: 42, Passed: 42, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 681, Passed: 662, Errors: 0, Failures: 2, Inconclusive: 0
Tests run: 248, Passed: 239, Errors: 0, Failures: 3, Inconclusive: 0
Tests run: 72, Passed: 72, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 5, Passed: 5, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 175, Passed: 173, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2, Passed: 2, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 169, Passed: 166, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 990, Passed: 928, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 146, Passed: 146, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 290, Passed: 289, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 737, Passed: 727, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 456, Passed: 456, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 36, Passed: 35, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 37, Passed: 37, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 315, Passed: 313, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 303, Passed: 298, Errors: 0, Failures: 1, Inconclusive: 0
Tests run: 303, Passed: 298, Errors: 1, Failures: 0, Inconclusive: 0
Tests run: 375, Passed: 375, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 164, Passed: 164, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 207, Passed: 206, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 9, Passed: 9, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 256, Passed: 237, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 35, Passed: 35, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 16, Errors: 103, Failures: 1, Inconclusive: 0
Tests run: 1422, Passed: 1422, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 65, Passed: 65, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 3, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 6, Passed: 6, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2, Passed: 2, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 14, Passed: 14, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 43, Passed: 43, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 17, Passed: 17, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 204, Passed: 203, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 67, Passed: 67, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
2600 test cases passed (99.54%)
3241 test cases passed (99.6%)
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
```
The majority of failures relate to:
* tailcalls where the eligibility is not as high on s390x as other platforms due to more restrictive conditions
* Debugger tests - sometimes causes the make check to hang on a single test
* There appear still to be a couple of endian issues in the tests 
* In a couple of tests short floating point numbers appear to be off on the last significant digit following a conversion from doubles. This may be due to a rounding flag and will be investigated

danmosemsft commented:
Poor @ViktorHofer you're getting a bunch of issues miscategorized infra. Not sure how to fix that.

ViktorHofer commented:
It makes me feel important 🦘😂

CoffeeFlux commented:
Maybe adjust the required confidence interval for the bot to actually apply a label? @danmosemsft

danmosemsft commented:
@maryamariyan is the bot owner. The confidence level seems to work pretty good except for this label. I guess we could hard code a higher threshold for this specific label. Maybe it will get better with future retraining.

CoffeeFlux commented:
Alternative, we could add a sanity check so that anything with from the mirror isn't getting a non-Mono label. No strong opinion there, the current situation just strikes me as a bit silly. 😄

danmosemsft commented:
That's a good idea as well. @maryamariyan

maryamariyan commented:
@CoffeeFlux @danmosemsft noted. I'll update the bot

nealef committed:
Major fixes to s390x ABI and add TAILCALL trace support

This pull request:
* Fixes the s390x ABI
  * For the case where structures are passed as parameters on the stack
  * Fixes the ```OP_S390_MOVE``` definition so register scheduling works correctly
  * Optimizes ```localloc``` generation
  * Fixes tailcall processing
    * Fixes register reloading
    * Adds ```OP_TAILLCALL_REG``` for s390x
    * Clarifies the conditions when tailcalls are supported on s390x
* Change function comment style to be doxygen-compatible
* Remove dead code
* Extract hardware facilities from CPU to enable use of newer opcodes
* Adds altstack support to exception processing for s390x
* Fixes a crash when using ```--trace``` with programs using tailcalls by providing its own icall method:
  * ```mono_trace_tail_method``` will use the callee and caller methods to report on the tailcall.
  * At the moment tailcalls are trace in ```mono_trace_leave_method``` where the second operand is expected to be a ```MonoProfileCallContext``` from which the return type and value are extracted which is meaningless for tail calls and worse may cause a SEGV.
  * The added call will use the parameters that are set up by ```mini_profiler_emit_tail_call```
* Fixes crashes in ```icall_System_Runtime_InteropServices_Marshal_copy_to_unmanaged``` and ```icall_System_Runtime_InteropServices_Marshal_copy_from_unmanaged```:
  * We now don't check the src/dest when the length field is zero so if either address is null it doesn't cause an exception
  * This is exhibited when an empty UTF16 string is converted to Byte[]. See this excerpt from ```WriteUTF16```in ```BlobBuilder.cs```:
    * For BigEndian systems the ```Encoding.Unicode.GetBytes(value);``` returns a null pointer for empty strings
    * ```WriteBytesUncheced``` invokes ```copy_from_unmanaged_fixed``` which crashes when given a null source pointer
```
            if (BitConverter.IsLittleEndian)
            {
                fixed (char* ptr = &value[0])
                {
                    WriteBytesUnchecked((byte*)ptr, value.Length * sizeof(char));
                }
            }
            else
            {
                byte[] bytes = Encoding.Unicode.GetBytes(value);
                fixed (byte* ptr = &bytes[0])
                {
                    WriteBytesUnchecked((byte*)ptr, bytes.Length);
                }
            }
```
The results of make check now show a higher pass rate:
```
22 test(s) passed, 0 test(s) did not pass.
617 test(s) passed, 71 test(s) did not pass.
216 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
9 test(s) passed, 0 test(s) did not pass.
9 test(s) passed, 0 test(s) did not pass.
4 test(s) passed, 0 test(s) did not pass.
4 test(s) passed, 0 test(s) did not pass.
1 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
2 test(s) passed, 0 test(s) did not pass.
6 test(s) passed, 0 test(s) did not pass.
Tests run: 10794, Passed: 10639, Errors: 2, Failures: 54, Inconclusive: 0
Tests run: 10, Passed: 10, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 634, Passed: 630, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 6401, Passed: 6260, Errors: 11, Failures: 6, Inconclusive: 0
Tests run: 2024, Passed: 1824, Errors: 0, Failures: 4, Inconclusive: 0
Tests run: 491, Passed: 466, Errors: 1, Failures: 0, Inconclusive: 0
Tests run: 200, Passed: 198, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1596, Passed: 1590, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 250, Passed: 250, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1821, Passed: 1804, Errors: 5, Failures: 2, Inconclusive: 0
Tests run: 24, Passed: 24, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 4, Passed: 4, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 89, Passed: 87, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 58, Passed: 58, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2405, Passed: 2054, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 100, Passed: 100, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 15, Passed: 15, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2891, Passed: 2774, Errors: 1, Failures: 6, Inconclusive: 0
Tests run: 22, Passed: 22, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 83, Passed: 33, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 61, Passed: 61, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 101, Passed: 97, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 4396, Passed: 4371, Errors: 1, Failures: 2, Inconclusive: 0
Tests run: 247, Passed: 244, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 256, Passed: 123, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 19, Passed: 19, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 5, Passed: 5, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 87, Passed: 26, Errors: 51, Failures: 7, Inconclusive: 0
Tests run: 42, Passed: 42, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 681, Passed: 662, Errors: 0, Failures: 2, Inconclusive: 0
Tests run: 248, Passed: 239, Errors: 0, Failures: 3, Inconclusive: 0
Tests run: 72, Passed: 72, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 5, Passed: 5, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 175, Passed: 173, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2, Passed: 2, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 169, Passed: 166, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 990, Passed: 928, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 146, Passed: 146, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 290, Passed: 289, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 737, Passed: 727, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 456, Passed: 456, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 36, Passed: 35, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 37, Passed: 37, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 315, Passed: 313, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 303, Passed: 298, Errors: 0, Failures: 1, Inconclusive: 0
Tests run: 303, Passed: 298, Errors: 1, Failures: 0, Inconclusive: 0
Tests run: 375, Passed: 375, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 164, Passed: 164, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 207, Passed: 206, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 9, Passed: 9, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 256, Passed: 237, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 35, Passed: 35, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 16, Errors: 103, Failures: 1, Inconclusive: 0
Tests run: 1422, Passed: 1422, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 65, Passed: 65, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 3, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 6, Passed: 6, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 2, Passed: 2, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 14, Passed: 14, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 43, Passed: 43, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 17, Passed: 17, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 1, Passed: 1, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 204, Passed: 203, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 67, Passed: 67, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 11, Passed: 11, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
2600 test cases passed (99.54%)
3241 test cases passed (99.6%)
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 21, Passed: 21, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 143, Passed: 142, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 454, Passed: 450, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 7, Passed: 7, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 229, Passed: 227, Errors: 0, Failures: 0, Inconclusive: 0
Tests run: 120, Passed: 120, Errors: 0, Failures: 0, Inconclusive: 0
```
The majority of failures relate to:
* tailcalls where the eligibility is not as high on s390x as other platforms due to more restrictive conditions
* Debugger tests - sometimes causes the make check to hang on a single test
* There appear still to be a couple of endian issues in the tests
* In a couple of tests short floating point numbers appear to be off on the last significant digit following a conversion from doubles. This may be due to a rounding flag and will be investigated
